pyomo/solver/base.py

```diff
@@ -0,0 +1,562 @@
+#  ___________________________________________________________________________
+#
+#  Pyomo: Python Optimization Modeling Objects
+#  Copyright (c) 2008-2022
+#  National Technology and Engineering Solutions of Sandia, LLC
+#  Under the terms of Contract DE-NA0003525 with National Technology and
+#  Engineering Solutions of Sandia, LLC, the U.S. Government retains certain
+#  rights in this software.
+#  This software is distributed under the 3-clause BSD License.
+#  ___________________________________________________________________________
+
+import abc
+import enum
+from typing import Sequence, Dict, Optional, Mapping, NoReturn, List, Tuple
+import os
+
+from pyomo.core.base.constraint import _GeneralConstraintData
+from pyomo.core.base.var import _GeneralVarData
+from pyomo.core.base.param import _ParamData
+from pyomo.core.base.block import _BlockData
+from pyomo.core.base.objective import _GeneralObjectiveData
+from pyomo.common.timing import HierarchicalTimer
+from pyomo.common.errors import ApplicationError
+from pyomo.opt.results.results_ import SolverResults as LegacySolverResults
+from pyomo.opt.results.solution import Solution as LegacySolution
+from pyomo.core.kernel.objective import minimize
+from pyomo.core.base import SymbolMap
+from pyomo.core.staleflag import StaleFlagManager
+from pyomo.solver.config import UpdateConfig
+from pyomo.solver.util import get_objective
+from pyomo.solver.results import (
+    Results,
+    legacy_solver_status_map,
+    legacy_termination_condition_map,
+    legacy_solution_status_map,
+)
+
+
+class SolverBase(abc.ABC):
+    """
```

Comment:
After thinking about this more, I think I like `opt.config.writer_config` and `opt.config.auto_updates`???

Reply:
I think `writer_config` is a bit too abstract / not approachable. I don't know if the use of persistent solvers would already indicate a power user, but if we intend for persistent solvers to be more widely usable, then it would be better to name options as clearly as possible. Most average users don't directly interact with the writer, right?

Reply:
`interface` is even more abstract... Also, for something like cyipopt that uses an nl writer, how do we distinguish writer options from cyipopt options?

Reply:
To clarify, there could be nl writer options, ipopt options (`solver_options`) and cyipopt specific options (I don't know that this is true, but I think it is a good example).

Reply:
Good discussion for the dev call.